watcher.go

```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
 
 // Size returns the number of views this watcher is watching.
 func (w *Watcher) Size() int {
-	w.depViewMapMx.Lock()
-	defer w.depViewMapMx.Unlock()
-	return len(w.depViewMap)
+	return w.tracker.viewCount()
 }
 
 // Remove-s the given dependency from the list and stops the
 // associated view. If a view for the given dependency does not exist, this
 // function will return false. If the view does exist, this function will return
 // true upon successful deletion.
 func (w *Watcher) remove(id string) bool {
-	w.depViewMapMx.Lock()
-	defer w.depViewMapMx.Unlock()
-
 	//log.Printf("[DEBUG] (watcher) removing %s", id)
 
 	defer w.cache.Delete(id)
-
-	if view, ok := w.depViewMap[id]; ok {
-		//log.Printf("[TRACE] (watcher) actually removing %s", id)
-		view.stop()
-		delete(w.depViewMap, id)
-		return true
-	}
-
-	//log.Printf("[TRACE] (watcher) %s did not exist, skipping", id)
-	return false
+	return w.tracker.remove(id)
 }
 
 // Watching determines if the given dependency (id) is being watched.
 func (w *Watcher) Watching(id string) bool {
-	w.depViewMapMx.Lock()
-	defer w.depViewMapMx.Unlock()
-
-	_, ok := w.depViewMap[id]
+	_, ok := w.tracker.lookup(id)
 	return ok
 }
 
-///////////
+// view is a convenience function for accessing stored views by id
+// note that dependency IDs and their corresponding view IDs are identical
+func (w *Watcher) view(id string) *view {
+	if v, ok := w.tracker.lookup(id); ok {
+		return v
+	}
+	return nil
+}
+
+///////////////////////////////////////////////////////////////////////////
 // internal structure used to track template <-> dependencies relationships
-type depmap map[string]map[string]struct{}
 
-func (dm depmap) add(k, k1 string) {
-	if _, ok := dm[k]; !ok {
-		dm[k] = make(map[string]struct{})
+func newTracker() *tracker {
+	return &tracker{
+		tracked:   make([]trackedPair, 0, 8),
+		views:     make(map[string]*view),
+		notifiers: make(map[string]Notifier),
 	}
-	dm[k][k1] = struct{}{}
 }
 
+// 1 view/notifier pair. Think many-2-many RDBMS table with annotations.
+type trackedPair struct {
+	// viewed: id of view watched, notified: id of external thing
+	// notified, client, consumer
+	view, notify string
+	// inUse flag gets off pre-render and back on at use
+	inUse bool
+}
+
+//	// removed flags that this pairing is no longer in use
+//	// fields get flagged for deletion, then removed on idle sweep
+//	removed bool
+
+func (tp trackedPair) used() trackedPair {
+	tp.inUse = true
+	return tp
+}
+
+func (tp trackedPair) refresh() trackedPair {
+	tp.inUse = false
+	return tp
+}
+
+type Notifier interface {
+	Notify(dep.Dependency)
+	ID() string
+}
+
+// If performance of looping through tracked gets to be to much build 2 indexes
+// of views/notifiers to their trackedPair entries and use that to accel lookups.
+// It will require updating though, and complicates things. So wait.
+
 type tracker struct {
-	sync.RWMutex
-	deps depmap
-	tpls depmap
+	sync.Mutex
```

Comment:
Good catch. I forget sometimes how embedding exposes its methods as public.

Reply:
I see why I left it. I tend not to worry about public/private on structs that are already internal/private.

I mean giving it a private name changes nothing in terms of access as anything with access to Watcher.tracker would also have access to Wather.tracker.mux. So changing it would not impact the ability to accidentally call locks. I guess it would make it a touch harder.

:thinking:

Reply:
@eikenb and I chatted about this offline and decided to keep the lock implementation as is. As mentioned in the previous comment, since the struct is private, public access to the lock is still prohibited. If the struct was public, we would definitely want to switch to a private lock. This implementation also makes calling the lock a little more elegant. But, if there's future concerns, we can update it later without any problems since it's private.